python/tchannel/statsd.py

```diff
@@ -0,0 +1,120 @@
+# THE SOFTWARE.
+
+from __future__ import absolute_import
+import re
+
+from .event import EventHook
+from .messages.error import ErrorMessage
+from .tornado.response import StatusCode
+
+
+class StatsdHook(EventHook):
+    """Collect Statsd information in the tchannel req/resp."""
+
+    def __init__(self, statsd):
+        """
+
+        :param statsd: instance of `StatsD <https://github.com/etsy/statsd>`
+        """
+        self._statsd = statsd
+
+    def before_send_request(self, request):
+        statsd_name = "tchannel.outbound.calls.sent"
+        key = common_prefix(statsd_name, request)
+
+        self._statsd.count(key, 1)
+
+    def after_receive_response(self, request, response):
+        if response.code == StatusCode.ok:
+            statsd_name = "tchannel.outbound.calls.success"
+        else:
+            statsd_name = "tchannel.outbound.calls.app-errors"
+        key = common_prefix(statsd_name, request)
+
+        self._statsd.count(key, 1)
+
+    def after_receive_system_error(self, request, error):
+        statsd_name = "tchannel.outbound.calls.system-errors"
+        prefix = common_prefix(statsd_name, request)
+        key = prefix + '.' + clean(
+            ErrorMessage.ERROR_CODES.get(error.code, None), 'type'
+        )
+
+        self._statsd.count(key, 1)
+
+    def after_receive_system_error_per_attempt(self, request, error):
+        statsd_name = "tchannel.outbound.calls.per-attempt.system-errors"
+        prefix = common_prefix(statsd_name, request)
+        key = prefix + '.' + clean(
+            ErrorMessage.ERROR_CODES.get(error.code, None), 'type'
+        )
+
+        self._statsd.count(key, 1)
+
+    def on_operational_error_per_attempt(self, request, error):
+        statsd_name = "tchannel.outbound.calls.per-attempt.operational-errors"
+        prefix = common_prefix(statsd_name, request)
+        key = prefix + '.' + clean(
+            ErrorMessage.ERROR_CODES.get(error.code, None), 'type'
+        )
+
+        self._statsd.count(key, 1)
+
+    def on_operational_error(self, request, error):
+        statsd_name = "tchannel.outbound.calls.operational-errors"
+
+        prefix = common_prefix(statsd_name, request)
+        key = prefix + '.' + clean(
+            ErrorMessage.ERROR_CODES.get(error.code, None), 'type'
+        )
+
+        self._statsd.count(key, 1)
+
+
+def extract_metadata(request):
+    service = request.headers.get('cn', None)
+    target_service = request.service
+    target_endpoint = request.endpoint
+
+    return (service, target_service, target_endpoint)
+
+
+def common_prefix(statsd_name, request):
+    (service, target_service, target_endpoint) = (
+        extract_metadata(request)
+    )
+
+    return '.'.join([statsd_name,
+                     clean(service, 'service'),
+                     clean(target_service, 'target-service'),
+                     clean(target_endpoint, 'target-endpoint')
+                     ])
+
+
+rexp = re.compile(r'[{}/\\:\s.]+')
```

Comment:
Use a better name (preferably all-caps) and put it at the top of the file after the imports, before the definitions.